framework/src/preference/preference_get.c

```diff
@@ -0,0 +1,296 @@
+ * Private Functions
+ ****************************************************************************/
+static int preference_fs_read_key(char *path, preference_data_t *data)
+{
+	int ret;
+	int type;
+
+	if (path == NULL || data == NULL) {
+		prefdbg("Invalid parameter\n");
+		return PREFERENCE_INVALID_PARAMETER;
+	}
+
+	int fd = open(path, O_RDONLY, 0666);
+	PREFERENCE_FREE(path);
+
+	if (fd < 0) {
+		if (errno == ENOENT) {
+			return PREFERENCE_KEY_NOT_EXIST;
+		}
+		return PREFERENCE_IO_ERROR;
+	}
+
+	/* Read the type of data value and Compare to the type of request */
+	ret = read(fd, (void *)&type, sizeof(int));
+	if (ret < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	} else if (type < 0 || data->type < 0 || data->type != type) {
+		prefdbg("Invalid type. request type:%d, read type:%d\n", type, data->type);
+		ret = PREFERENCE_INVALID_PARAMETER;
+		goto errout;
+	}
+
+	/* Read the len of data value */
+	ret = read(fd, (void *)&(data->len), sizeof(int));
+	if (ret < 0 || data->len < 0) {
+		ret = PREFERENCE_IO_ERROR;
+		goto errout;
+	}
+
+	/* Read the data value */
+	switch (data->type) {
+	case PREFERENCE_TYPE_INT:
+	case PREFERENCE_TYPE_BOOL:
+		ret = read(fd, (void *)&(data->value.i), data->len);
+		prefvdbg("[READ] int type %d\n", data->value.i);
+		break;
+	case PREFERENCE_TYPE_DOUBLE:
+		ret = read(fd, (void *)&(data->value.d), data->len);
+		prefvdbg("[READ] double type %f\n", data->value.d);
+		break;
+	case PREFERENCE_TYPE_STRING:
+		data->value.s = (char *)PREFERENCE_ALLOC(data->len);
+		if (data->value.s == NULL) {
+			ret = PREFERENCE_OUT_OF_MEMORY;
+			goto errout;
+		}
+		ret = read(fd, data->value.s, data->len);
+		prefvdbg("[READ] string type %s\n", data->value.s);
+		if (ret < 0) {
+			PREFERENCE_FREE(data->value.s);
+		}
+		break;
+	default:
+		prefdbg("Invalid type %d\n");
+		ret = PREFERENCE_INVALID_PARAMETER;
+	}
+
+	if (ret > 0) {
+		prefvdbg("Write key Success!\n");
+		ret = OK;
+	}
+errout:	
+	close(fd);
+
+	return ret;
+}
+
+static int preference_private_fs_read(char *key, preference_data_t *data)
+{
+	int ret;
+	char *path;
+
+	ret = preference_get_private_keypath(key, &path);
+	if (ret < 0) {
+		prefdbg("Failed to get preference path\n");
+		return ret;
+	}
+
+	return preference_fs_read_key(path, data);
+}
+
+static int preference_shared_fs_read(char *key_path, preference_data_t *data)
+{
+	int ret;
+	char *path;
+
+	if (key_path == NULL) {
+		prefdbg("Invalid parameter\n");
+		return PREFERENCE_INVALID_PARAMETER;
+	}
+
+	ret = preference_get_shared_keypath(key_path, &path);
+	if (ret < 0) {
+		prefdbg("Failed to get shared preference path\n");
+		return ret;
+	}
+
+	return preference_fs_read_key(path, data);
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+int preference_get_int(char *key, int *value)
+{
+	int ret;
+	preference_data_t result;
+
+	if (key == NULL) {
```

Comment:
Since `preference_fs_read_key` checks that `key is NULL, this code is unnecessary.
I think the function which directly uses `key` should check NULL.
So `preference_get_int` should give the responsibility of checking `key` to `preference_private_fs_read`, but check the `value`.
How do you think?

Reply:
right, I removed duplicated checking logics.